{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Project 1: Assembling Genomes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "\n",
    "   <div class=\"alert alert-block alert-danger\">\n",
    "    <center>Due: <b>Tuesday, 8 February, 4:59pm</b>.</center> \n",
    "   </div>\n",
    "   \n",
    "   <div class=\"alert alert-block alert-warning\">\n",
    "   <center>\n",
    "       <b>Collaboration and Resource Policy</b>\n",
    "    </center>\n",
    "    For this assignment, you are encouraged to work with one other person. Your team must satisfy the constraints mentioned in <a href=\"https://www.dropbox.com/s/g6z2xncwc4jsplp/csbio-class2-inked.pdf?dl=0\">Class 2</a>.\n",
    "    \n",
    "   1. You went to different high schools.\n",
    "   2. You and your partner have different answers to <em>at least one</em> of these questions:\n",
    "       - What country were you born in?\n",
    "       - Have you taken any biology courses at UVA?\n",
    "       - Have you taken cs3102?\n",
    "       - Have you taken cs4102?\n",
    "    \n",
    "We expect most students will have the best learning experience on this assignment by working with a partner, but if you prefer to work alone it is okay to do this assignment on your own.\n",
    "    \n",
    "You are permitted (actually _encouraged_) to discuss these problems with anyone you want, including other students in the class. If you do discuss the specific questions in the assignment with anyone other than your assignment partner and the course staff, though, you should list them in the _External resources used_ section below.\n",
    "    \n",
    "You are welcome to use any resources you want for this assignment, other than ones that would defeat the purpose of the assignment. This means you should not look at answers or code from any other students in the class (other than your collaboration with your partner), and if you find code that implements the problem you are being asked to do for the assignment, you should not use that code. You should document all external resource you use that are not part of the course materials in the _External resources used_ section below."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Team submitting this assignment:**  \n",
    "<div class=\"alert alert-success\">\n",
    "    <b><em>list each member of your team here, including both your name and UVA computing id\n",
    "        Colin Crowe (cbc6ghd) </em></b>\n",
    "</div>\n",
    "\n",
    "**External resources used:** \n",
    "<div class=\"alert alert-success\">\n",
    "<em>It is not necessary to list the course materials, but if you used any other resources, including discussing problems with students not on your team, list them here.</em>\n",
    "    \n",
    "    https://mathworld.wolfram.com/HamiltonianPath.html#:~:text=A%20Hamiltonian%20path%2C%20also%20called,cycle%20(or%20Hamiltonian%20cycle).\n",
    "    \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In this project, we will explore genome assembly â€” the process of determining the order of nucleotides from fragmented reads that are produced by sequencing machines. \n",
    "\n",
    "Genome assembly maps to a very well defined computer science problem, but can get quite complicated, as problems such as full sequence coverage, finding a good length for reads (the $k$ in $k$-mer), and sequencing errors present challenges for sequencing analysis and accuracy. For the required problems (ones everyone is expected to solve), you will be able to assume perfect coverage and no read errors; for the \"challenge\" problem (that is considered a bonus, and not something we expect everyone to be able to solve, and perhaps may not even be feasible for anyone to solve) you will not be able to rely on such assumptions."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    " <div class=\"alert alert-block alert-warning\">\n",
    "    \n",
    "<b>Submission</b>: You should work on this assignment by forking the provided Project 1 repository (you probably already did this, following the instructions posted at [https://computingbiology.github.io/project1](https://computingbiology.github.io/project1). Add the other teammate to as a collaborator with write access (also under the Settings tab) to this repository.\n",
    "    \n",
    "You should answer the questions and write your code in this Jupyter Notebook. (We don't expect you to need to use any external files or organize your code outside of the notebook, but if you do, make sure to put everything needed to run your code in your repository.) Parts where you are expected to provide and answer (which could be text that can be written in markdown format in the notebook or Python code that runs in the notebook) are marked in green.\n",
    "        \n",
    "When you are ready to submit the assignment, you should create a release of the version you are submitting ([github's directions for how to create a release](https://docs.github.com/en/repositories/releasing-projects-on-github/managing-releases-in-a-repository)) tagged as `submit`. After you've done this, send a message in slack to a channel that includes both team members (so the one sending this message should include the other team member) and all of the course staff (`dave`, `Hyun Jae Cho`, and `Anshuman Suri`) and a link to your repository (feel free to add any other useful comments if you want, but the link to your submission repository is sufficient)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Getting Started"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Install basic required packages, should be run only once. You may need to restart the jupyter python kernel (under the Kernel menu) after this. (You can execute this directly in the notebook but running the command below.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Requirement already satisfied: gvmagic in c:\\users\\fluffermuffin\\anaconda3\\lib\\site-packages (from -r requirements.txt (line 1)) (0.5)\n",
      "Requirement already satisfied: numpy in c:\\users\\fluffermuffin\\anaconda3\\lib\\site-packages (from -r requirements.txt (line 2)) (1.18.5)\n",
      "Requirement already satisfied: matplotlib in c:\\users\\fluffermuffin\\anaconda3\\lib\\site-packages (from -r requirements.txt (line 3)) (3.2.2)\n",
      "Requirement already satisfied: pydot in c:\\users\\fluffermuffin\\anaconda3\\lib\\site-packages (from -r requirements.txt (line 4)) (1.4.2)\n",
      "Requirement already satisfied: pyparsing!=2.0.4,!=2.1.2,!=2.1.6,>=2.0.1 in c:\\users\\fluffermuffin\\anaconda3\\lib\\site-packages (from matplotlib->-r requirements.txt (line 3)) (2.4.7)\n",
      "Requirement already satisfied: python-dateutil>=2.1 in c:\\users\\fluffermuffin\\anaconda3\\lib\\site-packages (from matplotlib->-r requirements.txt (line 3)) (2.8.1)\n",
      "Requirement already satisfied: kiwisolver>=1.0.1 in c:\\users\\fluffermuffin\\anaconda3\\lib\\site-packages (from matplotlib->-r requirements.txt (line 3)) (1.2.0)\n",
      "Requirement already satisfied: cycler>=0.10 in c:\\users\\fluffermuffin\\anaconda3\\lib\\site-packages (from matplotlib->-r requirements.txt (line 3)) (0.10.0)\n",
      "Requirement already satisfied: six>=1.5 in c:\\users\\fluffermuffin\\anaconda3\\lib\\site-packages (from python-dateutil>=2.1->matplotlib->-r requirements.txt (line 3)) (1.15.0)\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "%pip install -r requirements.txt"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "2. Make sure you have [graphviz](https://graphviz.org/) installed on your system. (On a Mac OS X, `brew install graphviz`. For other platforms, see [_https://graphviz.org/download/_](https://graphviz.org/download/).)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import collections\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "import utils"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 1: Assembling the TeleTubby Genome"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For this part, you're given reads generated while trying to sequence the DNA of a mysterious unknown organism with a _tiny_ genome. Some of the course staff is not sure what a TeleTubby is, but I've been assured that no cute creatures were harmed in producing this data (which was generated synthetically).\n",
    "\n",
    "By answering the following questions, you will learn how to assemble the original genome sequence from sequence reads.\n",
    "\n",
    "Sequencing data is often stored in the FASTQ file format, with is a simple ASCII format that is somewhat human-readable. \n",
    "\n",
    "In _TeleTubby.fastq_ (which is included in the repository you forked), you will find the data that was read from the TeleTubby genome. Each four lines of the file repeat the same pattern. For example,\n",
    "```\n",
    "@TeleTubby Genome: Project 1\n",
    "TAAAATGG\n",
    "+\n",
    "HAICDF5I\n",
    "```\n",
    "The first line contains the metadata that encodes the name of the read, the experiment type, the kind of sequencing machine used, etc. \n",
    "\n",
    "The second line is the sequence of bases that was read. \n",
    "\n",
    "The third line is just a placeholder.\n",
    "\n",
    "The fourth line is a sequence of base qualities that encodes the qualities for the corresponding bases in the sequence line. We will discuss and use this for Problem 2."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Read in the data from `TeleTubby.fastq`:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read sequence reads (error-free) from file\n",
    "sequence_reads, qualities = utils.read_fastq('TeleTubby.fastq')\n",
    "# print(qualities)\n",
    "# print(sequence_reads)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Melting Temperature\n",
    "\n",
    "The GC-content (or the ratio of G and C nucleotides) is related to the melting temperature of a DNA double helix structure. \n",
    "\n",
    "The following equation can be used to estimate the melting temperature (in degrees Celsius) of DNA for a particular sequence:\n",
    "\n",
    "\\begin{equation*}\n",
    "t_m = 64.9+0.41(\\%GC)-\\frac{500}{\\text{length of sequence}}\n",
    "\\end{equation*}\n",
    "\n",
    "As a reference, the human genome is known to have between 35%-60% GC-content. \n",
    "\n",
    "<div class=\"alert alert-success\">\n",
    "\n",
    "**Problem 1.** Calculate the melting temperature for the TeleTubby genome using the formula above. Assume that the sequence is 200 nucleotides long, and that the provided read data has equal coverage everywhere so the G-C distribution in the reads is very close to the actual G-C distribution for the genome. \n",
    "    </div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write code here for calculating the %GC content from the sequence_reads, and computing the estimated melting temperature.\n",
    "\n",
    "# first let's calculate %GC for each snippet\n",
    "gc = []\n",
    "for sequence in sequence_reads:\n",
    "    counter = 0\n",
    "    size = len(sequence)\n",
    "    for i in range(size):\n",
    "        char = sequence[i]\n",
    "        if char == 'G' or char == 'C':\n",
    "            counter += 1\n",
    "    gc.append((counter/size) * 100)\n",
    "\n",
    "#print(gc)\n",
    "\n",
    "# we need one number for the calculation, so let's average these together\n",
    "average = 0;\n",
    "for i in gc:\n",
    "    average += i;\n",
    "average = average / len(gc)\n",
    "\n",
    "#print(average)\n",
    "\n",
    "# cool, now let's just plug into the equation\n",
    "t = 0.41 * average\n",
    "t += 64.9\n",
    "t -= (500/300)\n",
    "\n",
    "# print(t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "my estimation for this genome's melting temperature is 84.74783845278726 degrees C.\n"
     ]
    }
   ],
   "source": [
    "# Print out temperature in Celsius\n",
    "print(\"my estimation for this genome's melting temperature is \" + str(t) + \" degrees C.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Interpreting Read Quality"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Phred33 quality scores are represented as the character with an ASCII code equal to its value + 33 (to make them easy to print alongside genome sequences). The Phred scores $Q$ are related to error probabilities $P$ as: $Q = -10 \\log_{10}(P)$. The table linked here provides a mapping from the Phred values provided in a fastq file and the $Q$ values: [Quality Score Encoding](https://support.illumina.com/help/BaseSpace_OLH_009008/Content/Source/Informatics/BS/QualityScoreEncoding_swBS.htm) \n",
    "\n",
    "<div class=\"alert alert-success\">\n",
    "    \n",
    "**Problem 2.** Using the Phred scores, identify the _least likely to be correct_ read in the TeleTubby.fastq data, which is the read with the lowest probability that it contains _no_ errors.\n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "the read with the highest probability of containing an error is the 14th read (GAGCCACC) with probability 0.04407109983095037 of having an error.\n"
     ]
    }
   ],
   "source": [
    "# Write your code for Problem 2 here. \n",
    "# It should output the read (nucleotide sequence) with the lowest probability of containing no errors.\n",
    "\n",
    "# ignore this - i'm just messing around here\n",
    "\n",
    "#test = ord('C') - 33\n",
    "#print(test)\n",
    "#test = test/-10\n",
    "#test = 10**test\n",
    "#print(test)\n",
    "\n",
    "# anyway,\n",
    "# for each read, we should convert the Q-scores to error probabilities,\n",
    "# do math to figure out the total probability,\n",
    "# then pick the read with the highest error\n",
    "\n",
    "# convert Q-scores to error probs\n",
    "errors = []\n",
    "for read in qualities:\n",
    "    read_errors = []\n",
    "    for i in range(len(read)):\n",
    "        Q = ord(read[i]) - 33\n",
    "        Q = Q/-10\n",
    "        Q = 10**Q\n",
    "        read_errors.append(Q)\n",
    "        \n",
    "    # do math to figure out the total probability\n",
    "    # the probability of there being an error in the sample is 1 minus the probability that there are no errors\n",
    "    # and the probability that there are no errors can just be multiplied together (assuming mutually exclusive)\n",
    "    count = 1\n",
    "    for j in read_errors:\n",
    "        no_error = 1 - j\n",
    "        count = count * no_error\n",
    "    errors.append(1 - count)\n",
    "\n",
    "# print(errors)\n",
    "\n",
    "# now we just want the read with the highest error\n",
    "\n",
    "#highest = max(errors) <-- doesn't actually work, we want to return which read had the highest one\n",
    "#print(highest)\n",
    "\n",
    "highest = 0\n",
    "the_one = 0\n",
    "for k in range(len(errors)):\n",
    "    if (errors[k] > highest):\n",
    "        highest = errors[k]\n",
    "        the_one = k\n",
    "        \n",
    "print(\"the read with the highest probability of containing an error is the \"\n",
    "      + str(the_one) + \"th read (\" + sequence_reads[the_one] + \") with probability \" + str(highest)\n",
    "     + \" of having an error.\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note: For the following problems, you can ignore the read quality scores, and safely assume there are no actual errors in the provided TeleTubby reads."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Frequency analysis\n",
    "\n",
    "Looking at repetitions in the sequence can be helpful in estimating the \"redudancy\" in the organisms. Eucaryote genomes have a lot of redundancy, while smaller organisms like bacteria have highly packed genomes. One heuristic to estimate this before actually performing the assembly could be looking at how often certain $k$-mers are repeated.\n",
    "\n",
    "<div class=\"alert alert-success\">\n",
    "    \n",
    "**Problem 3.**\n",
    "    Print out the 3 most frequent <em>k</em>-mers in the TeleTubby reads with their frequencies. (As in Problem 1, doing this from the reads only produces the same result as from the genome if the read coverage is equal everywhere, but it should be a close estimate.)\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "The most frequent k-mer is GA (appears 198 times)\n",
      "The second and third most frequent k-mers are AT (174 times) and AG (132 times)\n",
      "\n",
      "Excluding 2-mers, the most frequent k-mer is TGA (appears 90 times)\n",
      "The second and third most frequent k-mers are ATG (54 times) and TCG (38 times)\n"
     ]
    }
   ],
   "source": [
    "# Your code here\n",
    "\n",
    "# these snippets are just 8 characters long\n",
    "# so we can prolly just brute-force this without some fancy algorithm\n",
    "# (though finding some better way to do this would be p cool im just not smart enough for that lmao)\n",
    "\n",
    "mers = {} # data structure to hold our counts\n",
    "\n",
    "for i in range(8):\n",
    "    # we want to find all -mers of length 1, then 2, then 3, then 4...\n",
    "    i_mers = {}\n",
    "    i += 1 # range(8) starts at 0 and goes to 7, but we want it to start at 1 and go to 8\n",
    "\n",
    "    for sequence in sequence_reads:\n",
    "        for j in range(len(sequence)):\n",
    "            if j+i <= len(sequence): # our data gets messed up if we don't check for this (trust me)\n",
    "                the_mer = sequence[j:j+i] # grab a -mer of length i\n",
    "                if the_mer in i_mers.keys():\n",
    "                    i_mers[the_mer] += 1 # if we've already got it in the database then just increment it\n",
    "                else:\n",
    "                    i_mers[the_mer] = 1 # otherwise let's add it to the data structure\n",
    "\n",
    "    mers[str(i)] = i_mers # save these counts to the bigger database\n",
    "\n",
    "# print(mers['1']) <-- uncomment these if you wanna figure out wtf im doing up there\n",
    "# print(mers['2'])\n",
    "# print(mers['3'])\n",
    "\n",
    "# now let's figure out what the most frequent 3 are\n",
    "\n",
    "first = 0\n",
    "first_mer = ''\n",
    "second = 0\n",
    "second_mer = ''\n",
    "third = 0\n",
    "third_mer = ''\n",
    "\n",
    "for i in range(7):\n",
    "    i += 2 # somehow i don't think we should consider \"1-mers\" when deciding the most frequent k-mers... let's skip them\n",
    "    i_mers = mers[str(i)]\n",
    "    for j in i_mers:\n",
    "        if i_mers[j] > first:\n",
    "            first = i_mers[j]\n",
    "            first_mer = j\n",
    "        elif i_mers[j] > second:\n",
    "            second = i_mers[j]\n",
    "            second_mer = j\n",
    "        elif i_mers[j] > third:\n",
    "            third = i_mers[j]\n",
    "            third_mer = j\n",
    "            \n",
    "# print(first, first_mer)\n",
    "# print(second, second_mer)\n",
    "# print(third, third_mer)\n",
    "print(\"The most frequent k-mer is \" + first_mer + \" (appears \" + str(first) + \" times)\"\n",
    "      + \"\\nThe second and third most frequent k-mers are \" + second_mer + \" (\" + str(second) + \" times) and \"\n",
    "      + third_mer + \" (\" + str(third) + \" times)\")\n",
    "\n",
    "# this is still pretty unsatisfying tbh, \n",
    "# I could have told you that 2-mers would be the most common even without running this fancy program\n",
    "# so let's also print some larger k-mers\n",
    "\n",
    "# just copy-pasted from above\n",
    "first = 0\n",
    "first_mer = ''\n",
    "second = 0\n",
    "second_mer = ''\n",
    "third = 0\n",
    "third_mer = ''\n",
    "for i in range(6):\n",
    "    i += 3\n",
    "    i_mers = mers[str(i)]\n",
    "    for j in i_mers:\n",
    "        if i_mers[j] > first:\n",
    "            first = i_mers[j]\n",
    "            first_mer = j\n",
    "        elif i_mers[j] > second:\n",
    "            second = i_mers[j]\n",
    "            second_mer = j\n",
    "        elif i_mers[j] > third:\n",
    "            third = i_mers[j]\n",
    "            third_mer = j\n",
    "            \n",
    "print(\"\\nExcluding 2-mers, the most frequent k-mer is \" + first_mer + \" (appears \" + str(first) + \" times)\"\n",
    "      + \"\\nThe second and third most frequent k-mers are \" + second_mer + \" (\" + str(second) + \" times) and \"\n",
    "      + third_mer + \" (\" + str(third) + \" times)\")\n",
    "\n",
    "# neat that TCG is the third most common, I love Trading Card Games :)\n",
    "\n",
    "# print(mers['7'])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Greedy Assembly"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Given a set of sequence fragments, the objective of assembly can be viewed as finding the shortest sequence that contains all the fragments.\n",
    "\n",
    "One of the approaches to assemble the genome from the given reads is a greedy algorithm:\n",
    "\n",
    "```\n",
    "while (len(fragments) > 1):\n",
    "   calculate pairwise alignments of all pairs of fragments\n",
    "   merge the two fragments with the largest overlap\n",
    "genome = fragment[0] # the single remaining fragment is the genome\n",
    "```\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-success\">\n",
    "    \n",
    "**Problem 4 (a)**. What would the runtime be of this algorithm, given $n$ $k$-mer reads? (If you are not sure how to express your answer to this question, review [_Cost of Computation_](https://computingbiology.github.io/complexity/).)\n",
    "    </div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<i>Answer</i>: Calculating all pairwise allignments takes n^2, we do this n times, so n^3 runtime overall"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-success\">\n",
    "\n",
    "**Problem 4 (b)**. Is this algorithm guaranteed to find the correct genome? (A good answer will define what _correct_ means here, and explain why the algorithm is or is not guaranteed to find it.)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<i>Answer</i>: Well, we could define \"correct genome\" to be the one that was actually in the organism before we chopped it up and analyzed it. If we choose to define it like that the answer is clearly no. Like, let's say we had these snippets to work with:\n",
    "\n",
    "ATTA TAGA ACA AGA \n",
    "          \n",
    "Applying our algorithm, TAGA and AGA overlap in three places, so we can combine those:\n",
    "\n",
    "ATTA TAGA ACA\n",
    "\n",
    "Then, ATTA and TAGA overlap in two places:\n",
    "\n",
    "ATTAGA ACA\n",
    "\n",
    "Finally, ATTAGA and ACA overlap in one place. So in the end, we get this:\n",
    "\n",
    "ATTAGACA\n",
    "\n",
    "Or do we get this?\n",
    "\n",
    "ACATTAGA\n",
    "\n",
    "Our algorithm doesn't really help us here. Only one of these genes is what was actually in the organism before it got all chopped up, but our algorithm can't really figure out which one is more \"correct\" because they're the same length.\n",
    "\n",
    "And that's even assuming the full genome was this short to begin with - I started writing this section by cutting up this genome, actually:\n",
    "\n",
    "ATTAGAACAAGA\n",
    "\n",
    "So, no, if by \"correct\" we mean \"what the genome was before we cut it up\", this algorithm isn't guaranteed to give us that. But that's kind of boring, like, we said as much in class. Let's change \"correct\" to mean \"the shortest sequence that contains all the snippets\". Does this algorithm do that?\n",
    "\n",
    "If I remember 4102 correctly, we can show that this algorithm always produces the correct genome by doing a proof by cases and comparing it to a hypothetical, perfectly optimized algorithm. Every time this greedy algorithm decides to combine two fragments, there are two cases:\n",
    "\n",
    "<b>Case 1</b>: The optimal algorithm also combined those fragments the same way\n",
    "\n",
    "In this case, our greedy algorithm exactly matches the optimal algorithm.\n",
    "\n",
    "<b>Case 2</b>: The optimal algorithm did not combine those fragments that way\n",
    "\n",
    "In this case, the optimal algorithm must have combined those two fragments some other way. From here, there are two more possiblilties:\n",
    "\n",
    "<b>Case 2.1</b>: The optimal algorithm combined those fragments in a different way that led to the same amount of overlap\n",
    "\n",
    "In this case, the answers given by the optimal algorithm and our greedy algorithm are equally correct, since they have the same length.\n",
    "\n",
    "<b>Case 2.2</b>: The optimal algorithm combined those fragments in a different way that led to a different amount of overlap\n",
    "\n",
    "Our greedy algorithm already identified the combination that led to the highest overlap, so if the optimal algorithm chose to overlap them differnetly, then it must have overlapped them in a way that led to less overlap and, thus, a higher overall length. This means the so-called \"optimal\" solution wasn't actually optimal at all, and our greedy algorithm is better.\n",
    "\n",
    "So, in every case, our greedy algorithm is either just as correct as the optimal solution or more correct than that solution. Therefore, we can conclude that this greedy algorithm always produces the correct answer, so long as we choose to define \"correct\" this way."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Graph-based Assembly"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Graphs for genome assembly can be constructed in two ways:\n",
    "\n",
    "- Overlap graph: Processing $k-$mers as nodes, with $(k-1)-$mers as edges, and\n",
    "- de Bruijn graph: Processing $k-$mers as edges, with $(k-1)-$mers as nodes.\n",
    "\n",
    "A de Bruijn graph can be processed to find Euler paths, while Overlap graphs can be processed to find Hamiltonian paths. Both of these methods can be used reconstruct the original genome.\n",
    "\n",
    "<div class=\"alert alert-success\">\n",
    "    \n",
    "**Problem 5.**  Use one of these two techniques to reconstruct the TeleTubby genome from the provided sequence reads. \n",
    "    \n",
    "</div>\n",
    "    \n",
    "We have provided some template code below that may be helpful (but feel free to ignore this is you prefer)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read reads into graph\n",
    "\n",
    "def build_graph(sequence_reads):\n",
    "    \n",
    "    # ok look i'm not really sure what this skeleton code is doing (what's a gvmagic?)\n",
    "    # so uhhh let's just build this graph from scratch\n",
    "    # we'll encode it like this:\n",
    "    #   the graph will be a dictionary\n",
    "    #   keys are nodes\n",
    "    #   values are an array of edges (tuples)\n",
    "    #   each edge (tuple) has two values - the value of the edge (the k-mer) and the other node it's connected to\n",
    "    #   wait i lied that's so unnecessary nvm the edges will just be the node it's connected to\n",
    "    #   we already know what the value of the edge is because it's just the overlap between the two\n",
    "    \n",
    "    graph = {}\n",
    "    \n",
    "    # going with overlap here\n",
    "        \n",
    "    # the nodes should be k-mers\n",
    "    # so let's just grap all the samples since they all have the same length\n",
    "    for sequence in sequence_reads:\n",
    "        graph[sequence] = []\n",
    "            \n",
    "    # ok we've got our nodes, let's connect em' with edges\n",
    "    # ...this is going to be expensive isn't it -_-\n",
    "    # well it's at least n^2, so let's start there\n",
    "    for i in graph.keys():\n",
    "        for j in graph.keys():\n",
    "            # we want to link these two if the last 7 of the first one match the first 7 of the second\n",
    "            if i[1:] == j[0:-1]:\n",
    "                # add an edge fron i to j\n",
    "                graph[i].append(j)\n",
    "    \n",
    "        \n",
    "    \n",
    "    return graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [],
   "source": [
    "graph = build_graph(sequence_reads)\n",
    "\n",
    "# print(graph)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [],
   "source": [
    "# ok, now let's figure out this hamiltonian path\n",
    "# i used this to help me figure this thing out:\n",
    "# https://mathworld.wolfram.com/HamiltonianPath.html#:~:text=A%20Hamiltonian%20path%2C%20also%20called,cycle%20(or%20Hamiltonian%20cycle)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Main assembly algorithm\n",
    "\n",
    "# okay this is a mess lmao\n",
    "# ill try to add a gratuitous amount of comments to this thing\n",
    "\n",
    "def assemble_sequence(graph, path, already_visited, location): # recursive function, needs a lot of data passed down\n",
    "    num = len(graph.keys())\n",
    "\n",
    "    while(num != len(already_visited)): # don't stop until we visit every node\n",
    "        \n",
    "        nxt = graph[location] \n",
    "        viable = [] \n",
    "        \n",
    "        # find all the 'viable' nodes i.e. all the ones we haven't visited before and that we can jump to\n",
    "        for i in nxt: \n",
    "            if i not in already_visited:\n",
    "                viable.append(i)\n",
    "                \n",
    "        if len(viable) == 0:\n",
    "            # print(\"no hamiltonian path\") # print statements for debugging :P\n",
    "            return path\n",
    "        \n",
    "        # i don't think i had to do this part but i felt that if i could avoid using a recursive call, it's worth it\n",
    "        # esp since glancing at the graph it looks like a lot of these nodes just have one outgoing edge\n",
    "        elif len(viable) == 1:\n",
    "            location = viable[0]\n",
    "            path += location[7] # every new node we jump to adds one char to the path\n",
    "            already_visited.append(location)\n",
    "            # print(path) # more debugging - how do you think I noticed that CAGACCTG had no outgoing edges?\n",
    "            \n",
    "        # if there's more than one viable outgoing edge, we have to get ~funky~\n",
    "        else:\n",
    "            solutions = []\n",
    "            for j in viable: # set up a recursive call for each viable path\n",
    "                already_visited_local = already_visited # different calls want different lists of already_visited\n",
    "                path_local = path # ditto for path\n",
    "                location = j\n",
    "                path_local += location[7]\n",
    "                already_visited_local.append(location)\n",
    "                try_this = assemble_sequence(graph, path_local, already_visited_local, location)\n",
    "                solutions.append(try_this)\n",
    "                    \n",
    "            for k in solutions: # now that we've tried all the paths, pick the one that's the best and save it\n",
    "                if len(k) > len(path):\n",
    "                    path = k\n",
    "        \n",
    "    return path # ya done! \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CTCGAATCGTAAACGTGTGAATGTGTGATATGATGATTGATGAGTCCGATACGGGCGGATTCTGACGCAAAGCATCACCCCGTCGGGGTAAAAGAGCCTGACGACCGTGAGCATATGTCAGCATTTACCGGCGTTGCTCCCGACTTGTGCTCACCTGACAACATTAGGCAGGATAAGCTTATGAGATGACGCCCCCCCAGTGGTCAGATTTCCATCCCTGAGGATTTCGGTAGGACAATTTACCCATCGCGCGCCAGGGGTCGTAGGTAGAATTGCTCGGAGCCACCAGACCTG 294\n"
     ]
    }
   ],
   "source": [
    "# Output assembled sequence\n",
    "# Hint: Sequence is 200 nucleotides long\n",
    "\n",
    "# it doesn't matter where we start, since the cycle has to go through all the nodes\n",
    "# so uhh let's start at 0\n",
    "sequences = list(graph.keys())\n",
    "path = sequences[0]\n",
    "already_visited = [sequences[0]]\n",
    "location = sequences[0]\n",
    "\n",
    "# assmebled_seq = assemble_sequence(graph, path, already_visited, location)\n",
    "# print(assmebled_seq, len(assmebled_seq), len(sequences))\n",
    "# print(graph['CAGACCTG'])\n",
    "\n",
    "# okay, here's the thing\n",
    "# i misread some stuff up there\n",
    "# we want a hamiltonian *path*, not a hamiltonian *cycle*\n",
    "# why does this matter? CAGACCTG doesn't have any out going edges\n",
    "# so eventually this recursive thing will get stuck at CAGACCTG and return that no hamiltonian path exists\n",
    "# but a hamiltonian path *does* exist, just not a hamiltonian cycle\n",
    "\n",
    "# my solution?\n",
    "# ig we run it on every starting point and output the longest one\n",
    "# oh my god this is SUCH a MESS >_>\n",
    "\n",
    "possible_sequences = []\n",
    "for i in sequences:\n",
    "    path = i\n",
    "    already_visited = [i]\n",
    "    location = i\n",
    "    possible_sequences.append(assemble_sequence(graph, path, already_visited, location))\n",
    "    \n",
    "da_best = \"\"\n",
    "for i in possible_sequences:\n",
    "    if len(i) > len(da_best):\n",
    "        da_best = i\n",
    "\n",
    "print(da_best, len(da_best))\n",
    "\n",
    "# and sure enough, this one ends in CAGACCTG\n",
    "# god this question was a headache LOL"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-success\"> \n",
    "    \n",
    "**Problem 6.** Which of the two assembly methods did you use for problem 5, and why? (Hint: consider how the costs scale with the number of reads.)\n",
    "    \n",
    "</div>"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Part 2: Sequencing SARS-CoV-2 virus"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Let's move on from TeleTubbies to real-world organisms. For this problem, you'll assemble a genome for a variant of the SARS-CoV-2 virus. You're given reads from <i>actual</i> genome sequencing runs in the provided `SARS-CoV2.fastq` file. The file is based on the first Covid genome that was submitted on 5 January 2020: [https://www.ncbi.nlm.nih.gov/nuccore/NC_045512](https://www.ncbi.nlm.nih.gov/nuccore/NC_045512) But, you should be reconstructing a sequence from the provided reads without looking at the source sequence.\n",
    "\n",
    "<div class=\"alert alert-success\">\n",
    "    \n",
    "**Problem 7.**  Reconstruct the SATS-CoV2 genome from the provided sequence reads using $k=25$. \n",
    "    \n",
    "</div>\n",
    "\n",
    "You can re-use your implementation from Problem 5 and simply run it on the new data (depending on what you did for Problem 5, that might be enough to solve this problem). \n",
    "\n",
    "Print out your reconstructed sequence to a file `output.txt` (and add this file to the repo you submit). For this part, we will still assume that all the reads are error-free. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 148,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read sequence reads\n",
    "sequence_reads_covid, qualities_covid = utils.read_fastq('SARS-CoV2.fastq')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Read reads into graph\n",
    "graph_covid = build_graph(sequence_reads_covid)\n",
    "# print(graph_covid)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "22923 characters (1/29871)\n",
      "6305 characters (2/29871)\n",
      "14814 characters (3/29871)\n",
      "20995 characters (4/29871)\n",
      "29599 characters (5/29871)\n",
      "28619 characters (6/29871)\n",
      "24708 characters (7/29871)\n",
      "22807 characters (8/29871)\n",
      "8700 characters (9/29871)\n",
      "14657 characters (10/29871)\n",
      "2272 characters (11/29871)\n",
      "7967 characters (12/29871)\n",
      "2037 characters (13/29871)\n",
      "13952 characters (14/29871)\n",
      "15953 characters (15/29871)\n",
      "17503 characters (16/29871)\n",
      "15905 characters (17/29871)\n",
      "22382 characters (18/29871)\n",
      "18573 characters (19/29871)\n",
      "19675 characters (20/29871)\n",
      "6718 characters (21/29871)\n",
      "10767 characters (22/29871)\n",
      "4229 characters (23/29871)\n",
      "4348 characters (24/29871)\n",
      "27549 characters (25/29871)\n",
      "374 characters (26/29871)\n",
      "20556 characters (27/29871)\n",
      "13041 characters (28/29871)\n",
      "25080 characters (29/29871)\n",
      "24109 characters (30/29871)\n",
      "2504 characters (31/29871)\n",
      "16799 characters (32/29871)\n",
      "14969 characters (33/29871)\n",
      "13568 characters (34/29871)\n",
      "3346 characters (35/29871)\n",
      "29776 characters (36/29871)\n",
      "8723 characters (37/29871)\n",
      "28245 characters (38/29871)\n",
      "23316 characters (39/29871)\n",
      "27455 characters (40/29871)\n",
      "26416 characters (41/29871)\n",
      "2548 characters (42/29871)\n",
      "12574 characters (43/29871)\n",
      "27735 characters (44/29871)\n",
      "11131 characters (45/29871)\n",
      "4049 characters (46/29871)\n",
      "6456 characters (47/29871)\n",
      "19097 characters (48/29871)\n",
      "11244 characters (49/29871)\n",
      "6739 characters (50/29871)\n",
      "5063 characters (51/29871)\n",
      "17320 characters (52/29871)\n",
      "16159 characters (53/29871)\n",
      "19548 characters (54/29871)\n",
      "20335 characters (55/29871)\n",
      "11593 characters (56/29871)\n",
      "24113 characters (57/29871)\n",
      "8421 characters (58/29871)\n",
      "7253 characters (59/29871)\n",
      "29828 characters (60/29871)\n",
      "12527 characters (61/29871)\n",
      "7256 characters (62/29871)\n",
      "16400 characters (63/29871)\n",
      "12923 characters (64/29871)\n",
      "9159 characters (65/29871)\n",
      "16043 characters (66/29871)\n",
      "11152 characters (67/29871)\n",
      "14380 characters (68/29871)\n",
      "21353 characters (69/29871)\n",
      "6107 characters (70/29871)\n",
      "11292 characters (71/29871)\n",
      "19961 characters (72/29871)\n",
      "3742 characters (73/29871)\n",
      "25454 characters (74/29871)\n",
      "15399 characters (75/29871)\n",
      "24632 characters (76/29871)\n",
      "18686 characters (77/29871)\n",
      "26212 characters (78/29871)\n",
      "18863 characters (79/29871)\n",
      "17827 characters (80/29871)\n",
      "4239 characters (81/29871)\n",
      "24537 characters (82/29871)\n",
      "19531 characters (83/29871)\n",
      "7291 characters (84/29871)\n",
      "13273 characters (85/29871)\n",
      "13222 characters (86/29871)\n",
      "558 characters (87/29871)\n",
      "17044 characters (88/29871)\n",
      "656 characters (89/29871)\n",
      "11356 characters (90/29871)\n",
      "25803 characters (91/29871)\n",
      "13643 characters (92/29871)\n",
      "25271 characters (93/29871)\n",
      "28281 characters (94/29871)\n",
      "15321 characters (95/29871)\n",
      "15595 characters (96/29871)\n",
      "5133 characters (97/29871)\n",
      "10355 characters (98/29871)\n",
      "23964 characters (99/29871)\n",
      "16551 characters (100/29871)\n",
      "28097 characters (101/29871)\n",
      "21973 characters (102/29871)\n",
      "1606 characters (103/29871)\n",
      "13735 characters (104/29871)\n",
      "9959 characters (105/29871)\n",
      "1916 characters (106/29871)\n",
      "8898 characters (107/29871)\n",
      "498 characters (108/29871)\n",
      "19678 characters (109/29871)\n",
      "11579 characters (110/29871)\n",
      "19073 characters (111/29871)\n",
      "11555 characters (112/29871)\n",
      "17593 characters (113/29871)\n",
      "9742 characters (114/29871)\n",
      "24925 characters (115/29871)\n",
      "26430 characters (116/29871)\n",
      "21253 characters (117/29871)\n",
      "26769 characters (118/29871)\n",
      "8351 characters (119/29871)\n",
      "10547 characters (120/29871)\n",
      "9900 characters (121/29871)\n",
      "20870 characters (122/29871)\n",
      "4924 characters (123/29871)\n",
      "24973 characters (124/29871)\n",
      "1648 characters (125/29871)\n",
      "5177 characters (126/29871)\n",
      "7666 characters (127/29871)\n",
      "631 characters (128/29871)\n",
      "9905 characters (129/29871)\n",
      "24660 characters (130/29871)\n",
      "24317 characters (131/29871)\n",
      "19661 characters (132/29871)\n",
      "17303 characters (133/29871)\n",
      "21474 characters (134/29871)\n",
      "28110 characters (135/29871)\n",
      "20904 characters (136/29871)\n",
      "12599 characters (137/29871)\n",
      "18946 characters (138/29871)\n",
      "17071 characters (139/29871)\n",
      "21002 characters (140/29871)\n",
      "15978 characters (141/29871)\n",
      "16630 characters (142/29871)\n",
      "14481 characters (143/29871)\n",
      "16521 characters (144/29871)\n",
      "19265 characters (145/29871)\n",
      "22055 characters (146/29871)\n",
      "22267 characters (147/29871)\n",
      "25308 characters (148/29871)\n",
      "7032 characters (149/29871)\n",
      "27800 characters (150/29871)\n",
      "7509 characters (151/29871)\n",
      "13845 characters (152/29871)\n",
      "18695 characters (153/29871)\n",
      "12439 characters (154/29871)\n",
      "15281 characters (155/29871)\n",
      "24238 characters (156/29871)\n",
      "6674 characters (157/29871)\n",
      "286 characters (158/29871)\n",
      "14860 characters (159/29871)\n",
      "23374 characters (160/29871)\n",
      "3122 characters (161/29871)\n",
      "24653 characters (162/29871)\n",
      "17536 characters (163/29871)\n",
      "10366 characters (164/29871)\n",
      "11629 characters (165/29871)\n",
      "1941 characters (166/29871)\n",
      "18897 characters (167/29871)\n",
      "12220 characters (168/29871)\n",
      "18037 characters (169/29871)\n",
      "27352 characters (170/29871)\n",
      "11148 characters (171/29871)\n",
      "4422 characters (172/29871)\n",
      "23466 characters (173/29871)\n",
      "22404 characters (174/29871)\n",
      "5654 characters (175/29871)\n",
      "3543 characters (176/29871)\n",
      "3701 characters (177/29871)\n",
      "21959 characters (178/29871)\n",
      "10448 characters (179/29871)\n",
      "5193 characters (180/29871)\n",
      "25552 characters (181/29871)\n",
      "10915 characters (182/29871)\n",
      "22906 characters (183/29871)\n",
      "25948 characters (184/29871)\n",
      "26133 characters (185/29871)\n",
      "2798 characters (186/29871)\n",
      "24188 characters (187/29871)\n",
      "18525 characters (188/29871)\n",
      "18319 characters (189/29871)\n",
      "17957 characters (190/29871)\n",
      "4938 characters (191/29871)\n",
      "15161 characters (192/29871)\n",
      "18353 characters (193/29871)\n",
      "18688 characters (194/29871)\n",
      "1517 characters (195/29871)\n",
      "23387 characters (196/29871)\n",
      "5613 characters (197/29871)\n",
      "23391 characters (198/29871)\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-156-313acc7a59df>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m\u001b[0m\n\u001b[0;32m     15\u001b[0m     \u001b[0malready_visited\u001b[0m \u001b[1;33m=\u001b[0m \u001b[1;33m[\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     16\u001b[0m     \u001b[0mlocation\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0mi\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 17\u001b[1;33m     \u001b[0mpossible_sequences_covid\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0massemble_sequence\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mgraph_covid\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mpath\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0malready_visited\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mlocation\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     18\u001b[0m     \u001b[0mprint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mpossible_sequences_covid\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;33m-\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m \u001b[1;34m\"characters (\"\u001b[0m \u001b[1;33m+\u001b[0m \u001b[0mstr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mcounter\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m+\u001b[0m \u001b[1;34m\"/\"\u001b[0m \u001b[1;33m+\u001b[0m \u001b[0mstr\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mnum\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m+\u001b[0m \u001b[1;34m\")\"\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     19\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;32m<ipython-input-146-563a75dfc99f>\u001b[0m in \u001b[0;36massemble_sequence\u001b[1;34m(graph, path, already_visited, location)\u001b[0m\n\u001b[0;32m     15\u001b[0m         \u001b[1;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[1;32min\u001b[0m \u001b[0mnxt\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     16\u001b[0m             \u001b[1;32mif\u001b[0m \u001b[0mi\u001b[0m \u001b[1;32mnot\u001b[0m \u001b[1;32min\u001b[0m \u001b[0malready_visited\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 17\u001b[1;33m                 \u001b[0mviable\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mi\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     18\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     19\u001b[0m         \u001b[1;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mviable\u001b[0m\u001b[1;33m)\u001b[0m \u001b[1;33m==\u001b[0m \u001b[1;36m0\u001b[0m\u001b[1;33m:\u001b[0m\u001b[1;33m\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# Call main assembly algorithm\n",
    "# assmebled_covid_seq = assemble_sequence(nodes_covid, edges_covid)\n",
    "\n",
    "sequences_covid = list(graph_covid.keys())\n",
    "num = len(sequences_covid)\n",
    "counter = 0\n",
    "\n",
    "# ok i literally can't run this on every possible reading\n",
    "# it would just take too long\n",
    "# don't believe me? of course you do literally just look at it lmao\n",
    "# but if you actually don't believe me, run this\n",
    "# every time it finishes calculating 1 (one) run of assemble_sequence, it'll spit out how many characters its solution was\n",
    "# and how many more it has left to go\n",
    "# i left this thing on for about an hour while i took a shower and did some cleaning up\n",
    "# and it still wasn't finished\n",
    "# so uhhh at this point i think i just have to accept that the solution i made is bad\n",
    "# mission failed, we'll get em' next time i guess\n",
    "\n",
    "possible_sequences_covid = []\n",
    "for i in sequences_covid:\n",
    "    counter += 1\n",
    "    path = i\n",
    "    already_visited = [i]\n",
    "    location = i\n",
    "    possible_sequences_covid.append(assemble_sequence(graph_covid, path, already_visited, location))\n",
    "    print(len(possible_sequences_covid[-1]), \"characters (\" + str(counter) + \"/\" + str(num) + \")\")\n",
    "    \n",
    "# if you were curious, i started a timer on my phone\n",
    "# and determined that it took a minute and 23 seconds to complete 30 calls to assemble_sequence\n",
    "# that's about 2.8 seconds per call\n",
    "# which, at approx. 30,000 calls\n",
    "# would take 83,000 seconds\n",
    "# or just about 23 hours\n",
    "# ...this solution really is terrible, huh\n",
    "    \n",
    "da_best_covid = \"\"\n",
    "for i in possible_sequences_covid:\n",
    "    if len(i) > len(da_best_covid):\n",
    "        da_best_covid = i\n",
    "\n",
    "print(da_best_covid, len(da_best_covid))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Write assembled sequence to file\n",
    "with open(\"output.txt\", \"w\") as f:\n",
    "    f.write(\"solution runtime too slow to compute, see comments in notebook for details :(\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Part 3: Error-Aware Assembly (Challenge Problem)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "<div class=\"alert alert-warning\">\n",
    "    This problem is a \"Challenge Problem\". This means it is a problem of unknown difficulty that might be quite challenging (unlike the earlier problems, we don't have a reference solution for this one, or a clear idea how hard it might be). We do hope all students will at least attempt this and that more ambitious students will work hard to solve it and learn interesting things by the attempt (whether or not it is successful), but not get frustrated if you can't get to the desired answer.  As a \"Challenge Problem\" it means that you shouldn't be worried if you are not able to solve this, though, and you can get full expected credit on this assignment without answering it.\n",
    "</div>\n",
    "\n",
    "\n",
    "In the parts above, we assumed error-free reads while assembling $k$-mers. As much as we'd like that, actual reads can (and do) have errors, captured by their Phred scores. \n",
    "\n",
    "For this question, you're given raw, actual reads from sequencing runs. Download the reads from this file:\n",
    "https://sra-pub-sars-cov2.s3.amazonaws.com/sra-src/SRR11528307/ABS2-LN-R1_cleaned_paired.fastq.gz.  \n",
    "\n",
    "\n",
    "<div class=\"alert alert-success\">\n",
    "    \n",
    "**Problem 8 (Challenge).** Give the reads in the linked fastq file above, including their Phred33 quality score, assemble the most likely genome. Your solution should output the assembled sequence in `challenge.txt`. Provide a brief explantion of how your algorithm works and interesting things you learned in developing it.\n",
    "    \n",
    "</div>\n",
    "    \n",
    "This is an open-ended question. You are free to use any approach to deal with the issue. Make sure you provide your code, along with any assumptions you may have."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "_Write a description of your algorithm, and things you learned from working on this here._"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# implementation \n",
    "\n",
    "# so i've been going through a lot of personal stuff over the past couple weeks\n",
    "# it's part of the reason why i'm working alone on this, actually\n",
    "# i didn't think it'd be fair to try to work with someone else when i'm not sure if i'll be able to do my fair share\n",
    "\n",
    "# anyway point is i do think that this is an interesting question\n",
    "# and i'd like to attempt to answer it\n",
    "# but i just don't have it in me rn\n",
    "\n",
    "# still, my computer took like a billion years to finish calculating the covid genome (go figure)\n",
    "# so i had time to think over this question\n",
    "# and while i don't really have an answer to it\n",
    "# i do have some thoughts\n",
    "# i'll leave them here just to say i *did* think about it, and not much more\n",
    "\n",
    "\n",
    "# the first part seems to be figuring out what \"most likely genome\" even means\n",
    "# let's say that the most likely genome is the one that has the least probability of containing an error\n",
    "# simple, right? but this is still sort of weird to me\n",
    "# like, if we have two reads - ATTAGA and AGACTTA, what happens to the error probability when we overlap them on AGA?\n",
    "# i guess we make it so the error becomes the probability that either read had an error, right?\n",
    "# because they're both \"included\" in our final read\n",
    "# but then this is a problem, right?\n",
    "# because if we sequence this genome the same way we sequenced the COVID/TeleTubby one\n",
    "# then we'll include every read that was collected, regardless of how much error it had\n",
    "# so won't the error probability be equal among all the sequenced genes?\n",
    "# that doesn't seem very helpful\n",
    "\n",
    "# so clearly we need a different approach\n",
    "# something that i think might have potential is using a \"risk score\" or something\n",
    "# where Phred scores indicating high error probability increase the \"risk score\"\n",
    "# but this increase can be offset by high overlap between reads\n",
    "# so like if we have three reads - TGACCAAGTGAC, ACCAAGTGACCCG, and GACCAAGTGACC\n",
    "# if the ACCAAGTGAC part might has a lot of errors in one of the reads, that'd increase the \"risk score\"\n",
    "# but then the fact that we read it in three times indicates that it's probably part of the real genome\n",
    "# so we would decrease the \"risk score\" the more overlap we can find with questionable data\n",
    "\n",
    "\n",
    "# how could we implement that?\n",
    "# i think that it's best if we leave calculating the \"risk score\" up to the biologists\n",
    "# so let's let our sequencing function take in a couple functions as arguments that augment the risk score\n",
    "# maybe the two of them looks like this:\n",
    "\n",
    "def risk_score_increase(phread, risk):\n",
    "    risk += phread\n",
    "    return risk\n",
    "\n",
    "def risk_score_decrease(overlap, risk):\n",
    "    risk -= (overlap * 5)\n",
    "    return risk\n",
    "\n",
    "# and then whenever we find a way to combine two reads, we run both of these to modify the risk score\n",
    "# and at the end we output the shortest genome that incorporates all the reads and has the lowest risk score\n",
    "\n",
    "# or something like that, i dunno, that's all i got"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "   <div class=\"alert alert-block alert-danger\">\n",
    "    <center>\n",
    "        \n",
    " **Remember to follow the submission directions above to submit your assignment by 4:59pm on Tuesday, 8 February.**\n",
    "    \n",
    " </center>\n",
    " </div>"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
